docker-app/qfieldcloud/core/models.py

```diff
@@ -691,6 +692,64 @@ def active_users(self, period_since: datetime, period_until: datetime):
             is_staff=False,
         ).filter(Q(id__in=users_with_delta) | Q(id__in=users_with_jobs))
 
+    def list_active_users_jobs_deltas_count(self) -> List[Dict[str, Union[str, int]]]:
+        """Return a List of Dictionaries mapping user ids into the following values:
+        - user_id
+        - username
+        - number of jobs scheduled by the user
+        - number of deltas scheduled by the user
+        """
+        delta = timedelta(weeks=4)
+        now = datetime.now()
+        four_weeks_ago = now - delta
+
+        # Get the organization's projects & active users
+        organization_projects_pk: Set[str] = set(
+            Project.objects.filter(owner=self.pk).values_list("pk", flat=True)
+        )
+        organization_active_users_dict = dict(
+            self.active_users(four_weeks_ago, now).values_list("pk", "username")
+        )
```

Comment:
It seems super indirect to me to first use the acitve_users method in a first place here. To get the active users we already query all jobs and deltas by `created_at=in timerange` .

I would refactor this somehow to:

```
            Delta.objects.filter(
                project__in=self.projects.all(),
                created_at__gte=period_since,
                created_at__lte=period_until,
            )
       ...
        )
        users_with_jobs = (
            Job.objects.filter(
                project__in=self.projects.all(),
                created_at__gte=period_since,
                created_at__lte=period_until,
            )
```
A method / qs or maybe even custom manager, that returns all the jobs and deltas created in the requested period. Then use this qs in `def active_users` and add a distinct user query to it. And this qs in `def list_active_users_jobs_deltas_count` and add some count to it?

Reply:
You are onto something, yes, I can do without the `organization_active_users_dict`.